### ISSUE TITLE
The typespec of `iolist()` mentions `[]` as possible 'improper' terminator

### ISSUE THREAD

**Issue by Qqwy** [erlang/otp#6038]
**Describe the bug**

`iolist()` has been specified in the [typespec docs](https://www.erlang.org/doc/reference_manual/typespec.html) as being a shorthand for:

```erlang
iolist() :: maybe_improper_list(byte() | binary() | iolist(), binary() | [])
```

Why is the `[]` for the terminator there?
Isn't the fact that the list is allowed to be 'proper' already covered by the `maybe_improper` part of the name?

**Expected behavior**

```erlang
iolist() :: maybe_improper_list(byte() | binary() | iolist(), binary())
```

(c.f. https://github.com/elixir-lang/elixir/issues/11886)

**Comment by jhogberg** [comment]
No, the terminator doesn't implicitly include `[]` because that would make it impossible to express lists that are always terminated by non-nil terms. `maybe_improper_list(byte() | binary() | iolist(), binary())` would mean there's always a binary at the end.

The confusing name is there because the original authors didn't like creative uses of cons and wanted to discourage it, choosing to call the general list form `maybe_improper_list/2` instead of `list/2`.

**Comment by Qqwy** [comment]
@jhogberg Thanks for your reply!

I am however still a bit confused. I thought that `nonempty_improper_list(element_type(), terminator_type()` was the way to indicate a list which is always terminated by a non-nil term.
(And that by definition, an improper list can never be 'empty',  since the only empty kind of list would be nil).

In other words, that `maybe_improper_list(element_type(), terminator_type())` is syntactic sugar for `list(element_type()) | nonempty_improper_list(element_type(), terminator_type())`.

**Comment by jhogberg** [comment]
Nah, it's `list` and `nonempty_list` that are syntactic sugar for `maybe_improper_list(X, [])` and `nonempty_improper_list(X, [])`, respectively.

It gets confusing since `maybe_improper_list(X, [])` describes a proper list, but that's just the weird naming at work. I wish they would've called it `list/2` instead.

**Comment by Qqwy** [comment]
:+1: Thanks for clearing this up. Now I know that I need to handle this particular edge case in my runtime type-checking library.
